### DISCUSSION TITLE
Photos says that a lot of files are missing

### DISCUSSION THREAD

**Discussion by lgxz** [RhetTbull/osxphotos#1441] [Q&A]
Hello. 

This isn't a problem of osxphotos.  But I can't find a solution by Google/Claude/ChatGPT/Gemini. And I think that Apple wouldn't help me either. 

I have 100K photos.  I put all photo files in a partition (disk3s1) on a USB disk and the Photo Library in another partition (disk3s2).
So, Photo library doesn't contain photo files, only the references.

When I click some photos, Photos said: "Missing Files: Photos with unavailable original files cannot be opened. The original photo {file_name} is either offline or cannot be found. Click "Find original" to reconnect".

But when I click the "Find original" button, Photos will open a ChooseFolder dialog directly to the folder containing the "missing" file.  So, that means Photos indeed knows where the file is. 

I alse checked the row of the "missing" file  in ZASSET table of Photos.sqlite file.  Both ZDIRECTORY and ZFILENAME columns have correct value.

I tried repair photo library (Hold option+command), and it didn't work.

Now, I have to click random photos and then choose folder in the ChooseFolder dialog to fix the problem.  But I can't click thousands photos.

Does anyone know how to fix this problem? Or can osxphotos fix this problem?
Thanks very much!

<img width="262" alt="Screenshot 2024-03-09 at 23 17 22" src="https://github.com/RhetTbull/osxphotos/assets/858592/a12b1524-d554-40af-b1ab-ba36bfe84de4">

**Comment by oPromessa** [comment]
See #319 and #809

I have the same issue and never got around to solve it. I got it following conversion iPhoto to Photos and a few updates of Mac and Photos. Did not actually moved the pics location. I guess the volume ID changed which caused the problem. 

Learned to live with it.
- Recent imports don't have that problem. 
- I think I've used #809 or similar (SQL commmand over the DB -- risky. So make backups!) to fix some references on a Volume table with ids to the disk "Macintosh HD" which changed, following some Mac upgrades change the disks names I guess. 
- As it finds one pic in one folder it then stops asking for pics in that folder. 
- I have 100k pics and I still get that error on some older imported pics. 
- it's particularly annoying with videos as you don't see their thumbnail while browsing the media in Photos. 
- I guess it also impacts Face recognition somehow as Photos still notifies me it has processed 25k and 75k to go... but it's stuck. 

I've built an AppleScript -- but it's a bit cumbersome at this time -- to automate and go through the manual steps:
- Select a pic
- Check for find original
- hit ok on the folder
- way for a long time: it takes Photos like 60-90 seconds to "find the pic"
- go next pic
Did it for a few thousands. But not all. 

A full export with all metadata by `osxphotos` and create a new library and import might cleanup things. It's in the plans... but did not yet got around to do it!

In my case `osxphotos` is not affected as it finds and exports all photos. 


**update:** see also [this issue on RepairPhotosBookmarks](https://github.com/RhetTbull/RepairPhotosBookmarks/issues/2)

**Reply by RhetTbull** [reply]
See @oPromessa's answer. Bottom line: if photos are stored on an external disk and the disk is changed or moved to a new Mac, Photos will lose the references to the original files if you used referenced files instead of "copy items to the Photos library". Even though the path is present in the database, Photos uses a security-scoped bookmark to store the location of the file and this breaks if you move the files. There's not fix as far as I know. I made a [script](https://github.com/RhetTbull/RepairPhotosBookmarks) to fix these but it's not been updated in a while and I can't guarantee it will work on modern macOS.

**Reply by lgxz** [reply]
Thanks. Could you please share the AppleScript? 

I put all my photos in a directory structure as "year/month/day/".  I think I can pick up one photo each day. 
So the script could fix all photos in a year by running 365 times.

But there's still a lot of work because my photos are from 2002 to 2024.

**Reply by lgxz** [reply]
Thanks. I am running the script now. 

Photo pops up a lot of "import error" dialog.  But the script is still running.
<img width="640" alt="Screenshot 2024-03-10 at 13 32 14" src="https://github.com/RhetTbull/osxphotos/assets/858592/848bf3cf-4368-4a62-b211-19f44360f183">

**Reply by RhetTbull** [reply]
I suggest you search for one of those files in your directory to see what they are -- are they valid jpg files?

**Reply by lgxz** [reply]
Yes, they are valid.
I searched some of them on terminal and open them by the "open" command without any problem.

**Reply by lgxz** [reply]
By the way, the repair_photos_bookmarks.py didn't report any problem.
So I guess it was Photo's problem.

**Reply by lgxz** [reply]
Now repare_photos_bookmarks.py throws an exception and quits:

> line 351, in move_aae_file_if_it_exists
>   os.rename(aaepath, newpath)
> OSError: [Errno 30] Read-only file system: '/Volumes/Photo/Data/2019/06/29/10471900.aae' -> '/Volumes/Photo/Data/2019/06/29/10471900.aae.bak'

The partition I stored all my photos was mounted read-only. And all files and dirs are read-only and "setflags uchg".
 
And, I guess there are hundreds of "import error" dialog in Photos App so it is unusable. 
I have to click the "ok" button again and again.

**Reply by RhetTbull** [reply]
The repair bookmarks script has only been tested in a few use cases and not on latest version of macOS. I do not have time at the moment to look into this. Another option would be to import all the photos into a brand new library then use `osxphotos sync` command to copy the metadata (albums, keywords, etc) from the old library to the new one.
